processing/src/main/java/io/druid/segment/filter/ExpressionFilter.java

```diff
@@ -0,0 +1,127 @@
+import io.druid.query.expression.ExprUtils;
+import io.druid.query.filter.BitmapIndexSelector;
+import io.druid.query.filter.Filter;
+import io.druid.query.filter.ValueMatcher;
+import io.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import io.druid.segment.ColumnSelector;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.LongColumnSelector;
+import io.druid.segment.virtual.ExpressionSelectors;
+
+import java.util.Set;
+
+public class ExpressionFilter implements Filter
+{
+  private final Expr expr;
+  private final Set<String> requiredBindings;
+
+  public ExpressionFilter(final Expr expr)
+  {
+    this.expr = expr;
+    this.requiredBindings = ImmutableSet.copyOf(Parser.findRequiredBindings(expr));
+  }
+
+  @Override
+  public ValueMatcher makeMatcher(final ColumnSelectorFactory factory)
+  {
+    final LongColumnSelector selector = ExpressionSelectors.makeLongColumnSelector(factory, expr, 0L);
+    return new ValueMatcher()
+    {
+      @Override
+      public boolean matches()
+      {
+        return Evals.asBoolean(selector.get());
+      }
+
+      @Override
+      public void inspectRuntimeShape(final RuntimeShapeInspector inspector)
+      {
+        inspector.visit("selector", selector);
+      }
+    };
+  }
+
+  @Override
+  public boolean supportsBitmapIndex(final BitmapIndexSelector selector)
+  {
+    if (requiredBindings.isEmpty()) {
+      // Constant expression.
+      return true;
+    } else if (requiredBindings.size() == 1) {
+      // Single-column expression. We can use bitmap indexes if this column has an index and does not have
+      // multiple values. The lack of multiple values is important because expression filters treat multi-value
+      // arrays as nulls, which doesn't permit index based filtering.
+      final String column = Iterables.getOnlyElement(requiredBindings);
+      return selector.getBitmapIndex(column) != null && !selector.hasMultipleValues(column);
+    } else {
+      // Multi-column expression.
+      return false;
+    }
+  }
+
+  @Override
+  public <T> T getBitmapResult(final BitmapIndexSelector selector, final BitmapResultFactory<T> bitmapResultFactory)
+  {
+    if (requiredBindings.isEmpty()) {
+      // Constant expression.
+      if (expr.eval(ExprUtils.nilBindings()).asBoolean()) {
+        return bitmapResultFactory.wrapAllTrue(Filters.allTrue(selector));
+      } else {
+        return bitmapResultFactory.wrapAllFalse(Filters.allFalse(selector));
+      }
+    } else {
+      // Can assume there's only one binding and it has a bitmap index, otherwise supportsBitmapIndex would have
+      // returned false and the caller should not have called us.
+      final String column = Iterables.getOnlyElement(requiredBindings);
+      return Filters.matchPredicate(
+          column,
+          selector,
+          bitmapResultFactory,
+          str -> expr.eval(name -> str).asBoolean() // Don't look at "name", it will always be the single column name.
```

Comment:
Can you add some more description on what "name" refers to here (relating to ObjectBinding), was a little confused when I first saw this